README.md

```diff
@@ -0,0 +1,114 @@
+implementation with eventloop integration ability.
+
+## xtm_create
+
+Allocation function, used to create instance of struct xtm_queue and return
+its pointer or NULL with the following possible errno: ENOMEM, EINVAL, EMFILE,
+ENFILE, ENODEV. Accepts the queue size as input, which must be a power of two.
+
+## xtm_delete
+
+Deallocation function, used to free queue and close its internal fds.
+Returns 0 on success. Otherwise -1 with errno set appropriately (as in
+close(2), since it implies close of fds)
+
+## xtm_msg_notify
+
+Function for queue consumer notification.
+Returns 0 on success. Otherwise -1 with errno set appropriately (as in
+write(2), since it implies write to internal fd)
+
+## xtm_msg_probe
+
+Returns 0 if queue has space. Otherwise -1 with errno set to ENOBUFS.
+
+## xtm_msg_count
+
+Function returns current message count in queue.
+
+## xtm_fun_dispatch
+
+Function puts message containing the function and its argument to the queue.
+If delayed == 0 notifies consumer.
+Returns 0 on success. Otherwise -1 with errno set appropriately (as in
+write(2), since it implies write to internal fd + ENOBUFS in case there
+is no space in queue)
+
+## xtm_fd
+
+Returns file descriptor, that should be watched to become readable. When it
+became readable, consumer should call one of the consumer functions:
+xtm_msg_recv or xtm_fun_invoke.
+
+## xtm_fun_invoke
+
+Function retrieves messages from the queue and calls functions contained
+in them. If flushed == 1, flushing queue pipe. Return count of invoked
+functions on success. Otherwise -1 with errno set appropriately (as in
+read(2), since it implies read to internal fd)
+
+## xtm_msg_send
+
+Function puts message to the queue. If delayed == 0 notifies consumer.
+Returns 0 on success. Otherwise -1 with errno set appropriately (as in
+write(2), since it implies write to internal fd + ENOBUFS in case there
+is no space in queue)
+
+## xtm_msg_recv
+
+Function gets up to count elements from the queue. If flushed == 1 flushing
+queue pipe. Return count of extracted messages on success. Otherwise -1 with
+errno set appropriately (as in read(2), since it implies read to internal fd)
+
+Examples
+--------
+
+Full example you can see in test repository, here is only simple example of
+using library.
+
+**Create xtm queue**
+
+queue = xtm_create(XTM_QUEUE_SIZE);
+
+**Get xtm internal file descriptor to poll**
+
+fd = xtm_fd(queue);
+
+**Poll it in loop**
+
+fd_set readset;
+FD_ZERO(&readset);
+FD_SET(fd, &readset);
+select(fd + 1, &readset, NULL, NULL, NULL);
+
+**Process messages according invoke pattern**
+
+xtm_fun_invoke(queue, 1);
+
+**Process messages according receive pattern**
+
+void* msg[XTM_BATCH_MAX];
+int rc = xtm_msg_recv(queue, msg, XTM_BATCH_MAX, 1)
+
+**Destroy xtm queue, when it is no longer needed**
+
+xtm_delete(queue);
+
+**Send message to other thread, using queue created in other thread**
+
+while (xtm_msg_probe(queue) != 0)
+	;
```

Comment:
Is it the final and recommended way for waiting on the queue being writable again? How can I wait properly without the polling?

Reply:
Yes, i talk with @Mons about it. This is his answer:
"
We have discussed this for a long time. This requires either a callback interface or another file descriptor. At the moment, this is a strong overhead in api. And the call to probe in the loop is not needed: there is no place in the queue => the message fails (For H20 for example).
"

Reply:
Sorry, but this is hilarious: https://github.com/tarantool/httpng/blob/master/httpng/httpng.c#L402
So even before xtm is released, we already need to be able to wait on the queue being not full. And Mons is saying we won't need it while **we already need it**.

Reply:
> xtm_msg_probe called by producer thread! But file descriptor polled by consumer thread! So we need another pipe for producer thread notification

Yeah, I see now, indeed.

Reply:
> I talk about it with @alyapunov. He want to create another issue for it with discussion how to make it.

The problem is that xtm is hardly usable without this feature from what we see in httpng. All the usages use polling with `while true fiber.sleep(0)` which is strictly against Tarantool architecture - we don't have polling anywhere, we always rely on events on which we can subscribe and wait.

Another problem - whatever you will design might affect the API, which bothers me the most now.

Reply:
I think that if we can split a task into two - we should do it. Your suggestion is right, but It deserves a special issue. I'm also sure there are things to discuss in it.